src/application/controller/controller.c

```diff
@@ -1,2 +1,166 @@
+
+
 #include "application/controller/controller.h"
 
+#define GAIN_NUM 4
+
+QueueHandle_t internal_state_queue;
+QueueHandle_t output_queue;
+
+TickType_t timeout = pdMS_TO_TICKS(5);
+
+typedef struct {
+    float gain_arr[GAIN_NUM];
+    FILE *gain_table;
+    vector3d_t gain_k;
+    float gain_k_pre;
+    float reference_signal;
+} gain_table_entry_t;
+
+static controller_t controller_state = {0};
+static controller_input_t controller_input = {0};
+static controller_output_t controller_output = {0};
+static gain_table_entry_t gain_table_entry = {0};
+
+/*
+    TODO Send `canard_angle`, the desired canard angle (radians) to CAN
+*/
+static w_status_t controller_send_can(float canard_angle) {
+    // Build the CAN msg using [canard-specific canlib function to be defined
+    // later]. Send this to can handler module’s tx
+    return W_SUCCESS;
+}
+
+/**
+ * Initialize controller module
+ * Must be called before RTOS scheduler starts
+ * @return W_SUCCESS if initialization successful
+ */
+w_status_t controller_init(void) {
+    w_status_t init_status = W_SUCCESS;
+
+    // Create internal state queue (length = 1)
+    internal_state_queue =
+        xQueueCreate(1, sizeof(controller_input_t)); // estimator function to controller task
+    output_queue = xQueueCreate(1, sizeof(controller_output_t));
+
+    // check queue creation
+    if (internal_state_queue == NULL || output_queue == NULL) {
+        init_status = W_FAILURE;
+        log_text("controller", "queue creation failed");
+    }
+
+    // gain table file memory test
+    gain_table = fopen("file path", "r");
+
+    if (gain_table == NULL) {
+        log_text("controller", "gain table invalid");
+        init_status = W_FAILURE;
+    } else {
+        log_text("controller", "gain table file opened");
+    }
+
+    // Log initialization status
+    if (init_status == W_SUCCESS) {
+        log_text("controller", "initialization successful");
+
+    } else {
+        log_text("controller", "initialization failed");
+    }
+
+    // return w_status_t state
+    return init_status;
+}
+
+/**
+ * Update controller with new state data - called by state estimation module
+ * @param new_state Latest state estimate from state estimation
+ * @return W_FAILURE if validation/queueing fails
+ */
+w_status_t controller_update_inputs(controller_input_t *new_state) {
+    xQueueOverwrite(internal_state_queue, new_state); // overwrite internal queue
+    log_text("controller", "latest output received, timestamp %d", new_state->timestamp);
+    return W_SUCCESS;
+}
+
+/**
+ * Get most recent control output (cammanded angle and timestamp) - called by
+ * state estimation module
+ * @param output Pointer to store output -> type defined in controller
+ * @return W_FAILURE if no output available
+ */
+w_status_t controller_get_latest_output(controller_output_t *output) {
+    if (xQueuePeek(output_queue, output, timeout) == pdPASS) {
+        log_text("controller", "latest output returned");
```

Comment:
Again, redundant logging. In general, if something has to run every 5ms like controller/SE, we should never log nominal conditions